runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/stubs/IsolationChannel.java

```diff
@@ -0,0 +1,243 @@
+import java.util.logging.Logger;
+import javax.annotation.concurrent.GuardedBy;
+import org.apache.beam.sdk.annotations.Internal;
+import org.apache.beam.vendor.grpc.v1p60p1.io.grpc.CallOptions;
+import org.apache.beam.vendor.grpc.v1p60p1.io.grpc.ClientCall;
+import org.apache.beam.vendor.grpc.v1p60p1.io.grpc.ForwardingClientCall.SimpleForwardingClientCall;
+import org.apache.beam.vendor.grpc.v1p60p1.io.grpc.ForwardingClientCallListener.SimpleForwardingClientCallListener;
+import org.apache.beam.vendor.grpc.v1p60p1.io.grpc.ManagedChannel;
+import org.apache.beam.vendor.grpc.v1p60p1.io.grpc.Metadata;
+import org.apache.beam.vendor.grpc.v1p60p1.io.grpc.MethodDescriptor;
+import org.apache.beam.vendor.grpc.v1p60p1.io.grpc.Status;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.annotations.VisibleForTesting;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.base.Preconditions;
+
+/**
+ * A {@link ManagedChannel} that creates a dynamic # of cached channels to the same endpoint such
+ * that each active rpc has its own channel.
+ */
+@Internal
+class IsolationChannel extends ManagedChannel {
+  static final Logger LOG = Logger.getLogger(IsolationChannel.class.getName());
+
+  private final Supplier<ManagedChannel> channelFactory;
+
+  @GuardedBy("channelCache")
+  private final List<ManagedChannel> channelCache;
+
+  @GuardedBy("channelCache")
+  private final Set<ManagedChannel> usedChannels;
+
+  @GuardedBy("channelCache")
+  private boolean shutdownStarted = false;
+
+  private final String authority;
+
+  // Expected that supplier returns channels to the same endpoint with the same authority.
+  public static IsolationChannel create(Supplier<ManagedChannel> channelFactory) {
+    return new IsolationChannel(channelFactory);
+  }
+
+  @VisibleForTesting
+  IsolationChannel(Supplier<ManagedChannel> channelFactory) {
+    this.channelFactory = channelFactory;
+    this.channelCache = new ArrayList<>();
+    ManagedChannel channel = channelFactory.get();
+    this.authority = channel.authority();
+    this.channelCache.add(channel);
+    this.usedChannels = new HashSet<>();
+  }
+
+  @Override
+  public String authority() {
+    return authority;
+  }
+
+  /** Pick an unused channel from the set or create one and then create a {@link ClientCall}. */
+  @Override
+  public <ReqT, RespT> ClientCall<ReqT, RespT> newCall(
+      MethodDescriptor<ReqT, RespT> methodDescriptor, CallOptions callOptions) {
+    ManagedChannel channel = getChannel();
+    return new ReleasingClientCall<>(this, channel, channel.newCall(methodDescriptor, callOptions));
+  }
+
+  private ManagedChannel getChannel() {
+    synchronized (channelCache) {
+      if (!channelCache.isEmpty()) {
+        ManagedChannel result = channelCache.remove(channelCache.size() - 1);
+        usedChannels.add(result);
+        return result;
+      }
+    }
+    ManagedChannel result = channelFactory.get();
+    synchronized (channelCache) {
```

Comment:
I think there is a race where `isTerminated` could return false after returning true. 

1. Thread 1 is at `result = channelFactory.get()`
2. Thread 2 calls isTerminated gets True
3. Thread 1 makes progress and does `result.shutdown()` and exits getChannel
4. Thread2 calls isTerminated again and gets False

Is this okay?

Reply:
Changed so that we synchronize over channel creation and avoid creating new channels if we're shutting down, in that case we can use an existing channel.  One is guaranteed to exist from constructor.